### PR TITLE
6166: Migrate VirtualMap data from JasperDB to MerkleDb data sources

### PR DESCRIPTION
**Pull request by artemananiev**
(DRAFT) Fixes https://github.com/swirlds/swirlds-platform/issues/6166

Fix summary:

* The fix works only with a custom swirlds-platform branch, so compile time failures are expected at this moment
* New serialization classes (implement KeySerializer and ValueSerializer) for MerkleDb are added for storage, contracts storage, and unique tokens virtual maps
* A new bootstrap property is added, false by default: virtualdatasource.jasperdbToMerkledb. It indicates whether to perform JasperDB to MerkleDb migration for all virtual maps at node startup
* In `ServicesState`, when the state node is first created, it checks for the property and migrates all virtual map data for the 3 maps above. Migration is dumb, just iterating over the whole virtual tree, extracting keys and values from leaves, and putting them into the new map. Looking for ideas how to do it more efficiently
* After data is migrated to new virtual maps, the corresponding `ServicesState` child nodes are changed to the new maps

Testing done:

* Genesis start with property set to true. MerkleDb based virtual maps are created
* Genesis start with property set to false. Freeze, save state. Restart with the property set to true. JasperDB data sources are restored from the saved state and then closed after migration

Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

**Comment by artemananiev** [comment]
A few changes proposed by Cody and implemented in the latest commits in this PR:

* Instead of `addDeserializedChildren()`, VM migration is done in `migrate()` method in ServicesState
* To iterate over all entries in a virtual map, `VirtualMapMigration` utility class is used. Previously, it was done by iterating over all virtual leaf nodes
* When keys/values are added to new maps during migration, they pile up in virtual node cache. Since map may be large, caches may not be able to hold all the data in memory and need to be flushed to disk periodically. To implement that, maps are copied every 10_000 keys/values, which triggers background flushes

**Comment by artemananiev** [comment]
A small PR in swirlds/swirlds-platform repo required by this fix: https://github.com/swirlds/swirlds-platform/issues/6214

**Comment by artemananiev** [comment]
Additional testing using the saved state file from the repo:
* I ran ServicesMain with JasperDB, forced it to save the state (using “freeze” action), shut it down
* Then replaced the saved state with what’s in hedera-mono-service/src/test/resources/signedState.v0.27.4
* Started ServicesMain again with the new bootstrap property to migrate to MerkleDb set to true
* Verified virtual map entries are migrated, both ContractKey/ContractValue and VirtualBlobKey/VirtualBlobValue pairs

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ServicesState.java [hunk 1]
```diff
@@ -610,6 +627,73 @@ boolean shouldMigrateNfts() {
...
+        final int copyTargetMapEveryPuts = 10_000;
+        final AtomicInteger count = new AtomicInteger(copyTargetMapEveryPuts);
+        final AtomicReference<VirtualMap<K, V>> targetMapRef = new AtomicReference<>(target);
+        VirtualMapMigration.extractVirtualMapData(
+                AdHocThreadManager.getStaticThreadManager(),
+                source,
+                kvPair -> {
+                    final K key = kvPair.getKey();
+                    final V value = kvPair.getValue();
+                    log.info("Migrating " + key + " to " + value);
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ServicesState.java:683]: There are hundreds of millions of k/v pairs in some perfnet states, so we definitely don't want to log each migration.

But we might wrap this `extractVirtualMapData()` call [with `withLoggedDuration()`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/utils/MiscUtils.java#L868) to get a record of how long it takes.

**Comment by tinker-michaelj** [comment]
Everything looks very sensible to me! Just one little comment on logging overhead, and an apology for introducing some annoying merge conflicts [here](https://github.com/hashgraph/hedera-services/pull/4104) 😝.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ServicesState.java (hunk 1 above)
Comment (artemananiev) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ServicesState.java:683]: Yes, I noticed this line produces tons of output. It's just for debugging purpose for myself, so I'll remove it before merging.

**Comment by artemananiev** [comment]
@tinker-michaelj wrote:
> Everything looks very sensible to me! Just one little comment on logging overhead, and an apology for introducing some annoying merge conflicts [here](https://github.com/hashgraph/hedera-services/pull/4104) 😞.

The PR is synchronized with the latest `develop` branch changes. The next step is to add the two new maps to the fix, `accountsStore` and `tokenRelStore`. I'll do it shortly.

**Comment by artemananiev** [comment]
@tinker-michaelj I added accounts and token associations to the PR, please, have a look at your convenience.

**Comment by tinker-michaelj** [comment]
Will rework later.

**Pull request closed without merging**